### PR TITLE
Add Target, Dataset, Sample, Cohort

### PR DESCRIPTION
**Pull request by vladsaveliev**
Another huge PR, this time adding the abstract Target class, representing a stage target, following the pipeline's model. And three implementations: Sample, Dataset, Cohort. Dataset is somewhat tied to CPG dataset, so not sure if it needs some extra level abstraction for external users to decouple from CPG-specific stack vs. project details.

### CODE REVIEW

File: cpg_pipes/targets.py [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+        external_id: str | None = None,
+        participant_id: str | None = None,
+        sequencing_type: SequencingType | None = None,
+        meta: dict | None = None,
+        sex: Optional['Sex'] = None,
+        pedigree: Optional['PedigreeInfo'] = None,
+        alignment_input: AlignmentInput | None = None,
+    ) -> 'Sample':
+        """
+        Create a new sample and add it into the dataset.
```
Comment (lgruen) [inline cpg_pipes/targets.py:386]:
```suggestion
        Create a new sample and add it to the dataset.
```

File: cpg_pipes/targets.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+
+    def get_job_prefix(self) -> str:
+        """
+        Prefix job names.
+        """
+        return f'{self.dataset.name}/{self.id}: '
+
+
+@dataclass
+class PedigreeInfo:
```
Comment (lgruen) [inline cpg_pipes/targets.py:607]: Should we use the `peddy` classes here instead?

File: cpg_pipes/targets.py [hunk 3]
```diff
@@ -0,0 +1,639 @@
+"""
+Pipeline stage target and subclasses: Cohort, Dataset, Sample.
```
Comment (lgruen) [inline cpg_pipes/targets.py:2]: At the moment this also includes pedigree classes. Should those be moved to a separate module?

**Review by lgruen — APPROVED** [review]
Only added a few nits, but it would be good to get @illusional's review as well.

File: cpg_pipes/targets.py [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+
+    def alignment_inputs_hash(self) -> str:
+        """
+        Unique hash string of sample alignment inputs. Useful to decide
+        whether the analysis on the target needs to be rerun.
+        """
+        s = ' '.join(
+            sorted(
+                [
+                    str(s.alignment_input)
```
[the remark this answers is not in the document]
Comment (vladsaveliev) [reply cpg_pipes/targets.py:50]: Good idea! 
Added an issue. https://github.com/populationgenomics/production-pipelines/issues/38
Need to write tests for that feature anyway. There might be edge cases that I haven't thought through.

File: cpg_pipes/targets.py (hunk 3 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:2]: Thought about it a bit. We indeed don't need PedigreeInfo at all. Created an issue:
https://github.com/populationgenomics/production-pipelines/issues/39

File: cpg_pipes/targets.py (hunk 2 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:607]: Thought about it a bit. We indeed don't need PedigreeInfo at all. Created an issue:
https://github.com/populationgenomics/production-pipelines/issues/39

File: cpg_pipes/targets.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (illusional) [reply cpg_pipes/targets.py:51]: MD5 is usually stored in the metadata server where we are provided it - we also do check if we're provided an MD5.

File: cpg_pipes/targets.py [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+        Prefix job names.
+        """
+        raise NotImplementedError
+
+    def get_sequencing_type(self) -> SequencingType:
+        """
+        Sequencing type of samples. Will throw an error if more than
+        one sequencing type is found.
+        """
+        sequencing_types = set(s.sequencing_type for s in self.get_samples())
```
Comment (illusional) [inline cpg_pipes/targets.py:98]: As long as `SequencingType` is hashable, which I didn't think it was - maybe explicitly taking the string representation and then casting back to SequencingType.

File: cpg_pipes/targets.py [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+    def target_id(self) -> str:
+        """Unique target ID"""
+        return self.name
+
+    def get_datasets(self, only_active: bool = True) -> list['Dataset']:
+        """
+        Gets list of all datasets.
+        Include only "active" datasets (unless only_active is False)
+        """
+        return [
```
Comment (illusional) [inline cpg_pipes/targets.py:146]: What's an inactive dataset? I might be missing context so feel free to ignore if it's obvious with other context.

File: cpg_pipes/targets.py [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+    resolved to stack="seqr" and is_test=True, unless `namespace` is provided
+    explicitly.
+
+    Returns the stack id and a corrected namespace.
+    """
+    namespace = namespace or Namespace.MAIN
+    if name.endswith('-test'):
+        stack = name[: -len('-test')]
+        if namespace == Namespace.MAIN:
+            namespace = Namespace.TEST
```
Comment (illusional) [inline cpg_pipes/targets.py:235]: Why not just always set to TEST?

```suggestion
        namespace = Namespace.TEST
```

File: cpg_pipes/targets.py [hunk 8]
```diff
@@ -0,0 +1,639 @@
...
+        self,
+        analysis_dataset_name: str,
+        namespace: Namespace,
+        name: str | None = None,
+        storage_provider: StorageProvider | None = None,
+    ):
+        """
+        @param name: name of the cohort
+        @param analysis_dataset_name: deferring creation of the analysis
+        dataset in case if it is also a part of the cohort datasets.
```
[the remark this answers is not in the document]
Comment (vladsaveliev) [reply cpg_pipes/targets.py:123]: Sorry, that's a remnant from an older code. Cleaned that up!

File: cpg_pipes/targets.py [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+    def parse(sex: str | None) -> 'Sex':
+        """
+        Parse a string into a Sex object.
+        """
+        if sex:
+            if sex.lower() in ('m', 'male', '1'):
+                return Sex.MALE
+            if sex.lower() in ('f', 'female', '2'):
+                return Sex.FEMALE
+        return Sex.UNKNOWN
```
Comment (illusional) [inline cpg_pipes/targets.py:460]: I think UNKNOWN should be a specific case handled within the conditional, and unrecognisable values should be an exception:

```suggestion
            if sex.lower() in ('u', 'unknown', '0'):
                return Sex.UNKNOWN
            raise ValueError(f'Unrecognised sex value {sex}')
        return Sex.UNKNOWN
```

File: cpg_pipes/targets.py [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+        if sex:
+            self.pedigree = PedigreeInfo(
+                sample=self,
+                fam_id=self.participant_id,
+                sex=sex,
+            )
+        self.alignment_input: AlignmentInput | None = alignment_input
+
+    def __repr__(self):
+        return (
```
Comment (illusional) [inline cpg_pipes/targets.py:497]: Probably not a big deal, but could construct this a bit cleaner. This suggestion probably isn't formatted properly.

```suggestion
    values = {
        'participant': self._participant_id if self._participant_id else '',
        'forced': str(self.forced),
        'active': str(self.active),
        'meta': str(self.meta),
        'sequencing_type': str(self.sequencing_type.value),
        'alignment_input': self.alignment_input if self.alignment_input else ''
        'pedigree': self.pedigree if self.pedigree else ''
    }

    retval = f'Sample({self.dataset.name}/{self.id}'
    if self._external_id:
        retval += f'|{self._external_id}'
    return retval + ''.join(f', {k}={v}' for k, v in values.items())
```

File: cpg_pipes/targets.py [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+        ext_id = f'|{self._external_id}' if self._external_id else ''
+        return f'Sample({self.dataset.name}/{self.id}{ext_id}{ai_tag})'
+
+    @property
+    def participant_id(self) -> str:
+        """
+        Get ID of participant corresponding to this sample,
+        or substitute it with external ID.
+        """
+        return self._participant_id or self.external_id
```
Comment (illusional) [inline cpg_pipes/targets.py:534]: Should all samples that go through this process have a participant_id, should this be an assert?

**Review by illusional — COMMENTED** [review]
Looks good, nothing critical from me, just a few nitpicks as well.

File: cpg_pipes/targets.py (hunk 5 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:98]: Enums are hashable, seems to be working good!

```
>>> set([SequencingType.GENOME, SequencingType.EXOME])
{<SequencingType.EXOME: 'exome'>, <SequencingType.GENOME: 'genome'>}
>>> set([SequencingType.GENOME, SequencingType.GENOME])
{<SequencingType.GENOME: 'genome'>}
```

File: cpg_pipes/targets.py (hunk 6 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:146]: Base `Target` class defines `.active` field, which is inherited by all subclasses, so using it for consistency.
Also, in the pipelines logic, `active=False` is used in couple with `--skip-samples-with-missing-input` to flag targets (samples and datasets) that are not to be used in the pipeline:
https://github.com/populationgenomics/production-pipelines/blob/initial/cpg_pipes/pipeline/pipeline.py#L586-L587

File: cpg_pipes/targets.py (hunk 7 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:235]: Thanks!

File: cpg_pipes/targets.py (hunk 9 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:460]: Agree!

File: cpg_pipes/targets.py (hunk 10 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:497]: That looks much better!

File: cpg_pipes/targets.py (hunk 11 above)
Comment (vladsaveliev) [reply cpg_pipes/targets.py:534]: I don't want to force non-CPG users to use participant IDs. I think they should be able to use only one type of IDs if they want, and the code that expects participant IDs still should work
